File: python/utils/datetimeutils.py
### Date/time utilities
## DEPENDENCIES:
# OS: 
# Python: 

import datetime
import time

class DateTimeUtils():
    def __init__(self, utc_offset):
        self.__utc_offset = utc_offset
        
    # return the timestamp with the timezone offset applied
    def timezone(self, timestamp):
        return int(timestamp+self.__utc_offset*3600)

    # return an UTC timestamp from a local timezone timestamp
    def utc(self, timestamp):
        return int(timestamp-self.__utc_offset*3600)
        
    # return the now timestamp (in the current timezone)
    def now(self):
        return self.timezone(int(time.time()))

    # return yesterday's timestamp (in the local timezone)
    def yesterday(self):
        return self.now()-24*3600

    # return the last hour timestamp (in the local timezone)
    def last_hour(self):
        return self.now()-60*60
        
    # generate a given timestamp based on the input
    def get_timestamp(self, years, months, days, hours, minutes, seconds):
        date = datetime.datetime(years, months, days, hours, minutes, seconds, 0)
        return int((date - datetime.datetime(1970, 1, 1)).total_seconds())

    # return day start timestamp
    def day_start(self, timestamp):
        date = datetime.datetime.utcfromtimestamp(timestamp)
        return self.get_timestamp(date.year, date.month, date.day, 0, 0, 0)
        
    # return day end timestamp
    def day_end(self, timestamp):
        date = datetime.datetime.utcfromtimestamp(timestamp)
        return self.get_timestamp(date.year, date.month, date.day, 23, 59, 59)

    # return hour start timestamp
    def hour_start(self, timestamp):
        date = datetime.datetime.utcfromtimestamp(timestamp)
        return self.get_timestamp(date.year, date.month, date.day, date.hour, 0, 0)

    # return hour end timestamp
    def hour_end(self, timestamp):
        date = datetime.datetime.utcfromtimestamp(timestamp)
        return self.get_timestamp(date.year, date.month, date.day, date.hour, 59, 59)
        
    # return a timestamp as a human readable format
    def timestamp2date(self, timestamp):
        return datetime.datetime.utcfromtimestamp(self.utc(int(timestamp))).strftime('%Y-%m-%d %H:%M:%S')

File: javascript/sdk/utils/datetimeutils.js


class DateTimeUtils {
    constructor(utc_offset) {
        this.__utc_offset = utc_offset
    }
    
    // adjust the given timestamp to the local timezone
    timezone(timestamp) {
       return parseInt(timestamp+this.__utc_offset*3600)
    }

    // adjust the given timestamp to utc
    utc(timestamp) {
       return parseInt(timestamp-this.__utc_offset*3600)
    }
    
    // return the current timestamp, in the local timezone
    now() {
        return this.timezone(parseInt(Date.now()/1000), this.__utc_offset)
    }
    
    // format the provided timestamp for printing
    format_timestamp(timestamp=this.now()) {
        var date = new Date(this.utc(timestamp) * 1000)
        var day = date.getDate() < 10 ? "0"+date.getDate() : date.getDate()
        var month = (date.getMonth()+1) < 10 ? "0"+(date.getMonth()+1) : date.getMonth()+1
        var year = date.getFullYear()
        var hour = date.getHours() < 10 ? "0"+date.getHours() : date.getHours()
        var minute = date.getMinutes() < 10 ? "0"+date.getMinutes() : date.getMinutes()
        var second = date.getSeconds() < 10 ? "0"+date.getSeconds() : date.getSeconds()
        return day+"/"+month+"/"+year+" "+hour+":"+minute+":"+second
    }

    // return the difference between two timestamps in a human readable format
    timestamp_difference(date1, date2) {
        if (date1 == "" || date2 == "") return "N.A"
        // TODO: localization
        var seconds = Math.floor(Math.abs(date1-date2))
        var interval = Math.floor(seconds / 31536000)
        if (interval > 1) return String(parseInt(interval)) + " years ago"
        interval = Math.floor(seconds / 2592000)
        if (interval > 1) return String(parseInt(interval)) + " months ago"
        interval = Math.floor(seconds / 86400)
        if (interval > 1) return String(parseInt(interval)) + " days ago"
        interval = Math.floor(seconds / 3600)
        if (interval > 1) return String(parseInt(interval)) + " hours ago"
        interval = Math.floor(seconds / 60)
        if (interval > 1) return String(parseInt(interval)) + " minutes ago"
        return String(parseInt(Math.floor(seconds))) + " seconds ago"
    }
}
